stm32: dma: enable dma with peripheral using DMAMUX

FRASTM opened the pull request:
Enable peripheral to memory or memory to peripheral dma operations through the DMAMUX on STM32xx. 
It upgrades the dma feature, with the dmamux controller for some stm32 soc like stm32wbxx or stm32g4xx or stm32l4R5.
The DMAMUX acts as a programmable request source selector to give full dynamic DMA peripheral request mapping": the client of the dma just needs to set the request and choose the DMAMUX channel.
The periph-to/from-mem transfer allocation is no more fixed in a onfiguration table. 
During dma transfer with dmamux, the dma driver does not handle the Half Transfer IRQ (ht).
For more information refer to [STMicroelectronics AN5224](https://www.st.com/content/ccc/resource/technical/document/application_note/group1/c5/95/ed/03/0f/ec/48/37/DM00535045/files/DM00535045.pdf/jcr:content/translations/en.DM00535045.pdf).
Tested on nucleo_wb55rg for mem-to-mem and peripheral to/from memory transfers, when the client (spi) is configured ( PR #23157) and test applications are available (PR #23155)

This PR is required to fix https://github.com/zephyrproject-rtos/zephyr/issues/20825

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
add the sample application : dma/mux_req_gen 
It configures the dmamux to rig the dma transfer the request generator coming from the user button.
As the dmamux channel is not yet configurable, the EXTI line 13 is fixed.

FRASTM commented:
add USE_STM32_LL_DMAMUX definition in  modules/Kconfig.stm32
removing blank lines at end of 'tests/drivers/dma/mux_req_gen/testcase.yaml'

erwango commented on drivers/dma/dma_stm32_v2.c:
```diff
@@ -21,8 +21,12 @@ u32_t table_ll_stream[] = {
 	LL_DMA_CHANNEL_3,
 	LL_DMA_CHANNEL_4,
 	LL_DMA_CHANNEL_5,
-	LL_DMA_CHANNEL_6,
```
I think this line should not have been deleted.

FRASTM replied:
#if defined(LL_DMA_CHANNEL_6)
	LL_DMA_CHANNEL_6,
#endif /* LL_DMA_CHANNEL_6 */

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -23,10 +23,17 @@ config DMA_STM32_V1
 
 config DMA_STM32_V2
 	bool
-	depends on SOC_SERIES_STM32F0X || SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32L0X || SOC_SERIES_STM32L4X ||SOC_SERIES_STM32WBX
```
Miss a ">" at the end of the "Signed-off-by" for this commit.
Signed-off-by: Francois Ramu <francois.ramu@st.com

erwango replied:
Commit title (memory/periph) and message (memory to memory ) do not match

erwango commented on dts/bindings/dma/dmamux-controller.yaml:
```diff
@@ -0,0 +1,57 @@
+  - gen
+  - gen-config
+
+# channel: DMAMUX channel this value cannot exceed the <dmamux-channels>
```
Miss a "." before "This  value ..."

erwango commented on dts/bindings/dma/dmamux-controller.yaml:
```diff
@@ -0,0 +1,57 @@
+
+# channel: DMAMUX channel this value cannot exceed the <dmamux-channels>
+# channel-config: configuration of the selected DMAMUX channel request (DMAMUX_CxCR)
+# gen: DMAMUX request generator input resource ; this value is >=0 and cannot exceed <dmamux-gen-nb>
```
'this value is >=0 and cannot exceed <dmamux-gen-nb>'

Either > 0, either cannot exceed dmamux-gen-nb-1.
Reading below line, maybe "0" has a specific meaning? If yes, can you elaborate?

FRASTM replied:
gen: DMAMUX request generator input resource cannot exceed  `dmamux-gen-nb - 1`

erwango commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -18,6 +18,14 @@ properties:
+    st,mem2periph:
+      type: boolean
+      description: If the controller supports memory to pheripheral transfer
+
```
This addition looks not specific to DMAMUX, if true, put this in a dedicated commit (prior to dmamux ones if necessary for dmamux).
Also, existing "st,mem2mem" was currently used to assert support of mem2mem operation. 
Support of mem2periph and periph2mem were implicit, and assumed this way in the driver.
Adding  "st,periph2mem" and "st,mem2periph" is making support of these capability explicit, which, I guess, has impact on the driver. This impact has to be put in a dedicated commit, prior to pure DMAMUX related changes.

FRASTM replied:
We will keep st,mem2mem if the dma instance ONLY supports memory to memory transfer (ex. DMA2 of the stm32F4xx). For any other instance (all transfer directions are possible), the st,mem2mem is not set.

FRASTM replied:
for L4 series, the dma should not include the mem2mem (dt/arm/st/l4/stm32l4r5.dtsi)

FRASTM replied:
changed:  except for DMA V1, where mem2mem is a limitation

erwango commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -30,3 +38,8 @@ dma-cells:
   - features
+
+# channel: DMA channel
+# slot: DMA stream of the DMA channel for DMA V1 or MUX peripheral request for DMA V2
```
"MUX peripheral request for DMA V2"
Some V2 don't use DMAMUX, please rework the comment.

FRASTM replied:
slot: DMA stream of the DMA channel for DMA V1 or for DMA V2 with MUX peripheral request else NA

erwango reviewed (changes requested):
Reviewed the non dmamux specific part.

FRASTM commented:
change comment for description of `st,mem2mem` 
in dts/bindings/dma/st,stm32-dma.yaml

FRASTM commented:
fix Identity/Emails issues

FRASTM commented:
dmamux configuration for stm32L4R5

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -23,9 +23,10 @@ config DMA_STM32_V1
 
 config DMA_STM32_V2
 	bool
-	depends on SOC_SERIES_STM32F0X || SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32L0X || SOC_SERIES_STM32L4X
```
Ideally, this patch should be identical to commit from https://github.com/zephyrproject-rtos/zephyr/pull/20824
and this PR should be based on top and advertised as such.

FRASTM replied:
done, see commit 1423d4ed08abdff9125689d1ff2b48b3a1c38e9f

erwango commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -4,8 +4,11 @@
 #define STM32_DMA_CONFIG_PERIPH_ADDR_INC(config)	(config & 0x1 << 9)
 #define STM32_DMA_CONFIG_MEM_ADDR_INC(config)		(config & 0x1 << 10)
+#define STM32_DMA_CONFIG_PERIPH_DATA_SIZE(config)	(config & (0x3 << 11))
+#define STM32_DMA_CONFIG_MEM_DATA_SIZE(config)		(config & (0x3 << 13))
```
This should come with the equivalent yaml binding.

FRASTM replied:
moved, see commit 383b94d117c7cbe43a5892b2ff13b99955b97f52

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -11,14 +11,77 @@
+#include <drivers/dma.h>
+#include <drivers/clock_control.h>
+#include <clock_control/stm32_clock_control.h>
+
```
Some of these inclusions are redundant with below inclusions

FRASTM replied:
ok

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -11,14 +11,77 @@
+	}
+
+	/* then callback */
+	data->callback_arg = dmamux;
```
Already done line 54

FRASTM replied:
removed from line 54:
note that we could one callback for each type of error ( Syncr or request generator)

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -36,6 +99,24 @@ static void dma_stm32_clear_stream_irq(struct device *dev, u32_t id)
+#if defined(CONFIG_DMA_STM32_V2) && defined(CONFIG_DMAMUX_STM32)
+	/* also clear the corresponding dmamux irq */
+	struct device *dev_mux =
+		 device_get_binding((const char *)"DMAMUX_1");
```
Miss a comment to explain why this is valid on DMA_1 only

FRASTM replied:
no, with DMAMUX only
clearing the dma channel (stream) irq will also clear the corresponding dmamux channel irq

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -36,6 +99,24 @@ static void dma_stm32_clear_stream_irq(struct device *dev, u32_t id)
+					dev_mux->config->config_info;
+	/* retrieve the dma channel from the dmamux channel */
+	u32_t mux_id = ((strcmp(dev->config->name, "DMA_1") == 0) ? id :
+	 ((devmux_config->channel_nb / 2) + id));
```
This should be explained

FRASTM replied:
this formula is used to retrieve the dmamux channel from the dma instance + channel (stream)
put in a macro

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -62,7 +143,13 @@ static void dma_stm32_irq_handler(void *arg)
 	stream = &data->streams[id];
-	stream->busy = false;
+	/*
+	 * when using the dmamux, in periph2memm or mem2periph
```
If valid only for DMAMUX, should it be under "#ifdef CONFIG_DMAMUX_STM32"

FRASTM replied:
yes

FRASTM commented:
fix warning: parameter names (without types) in function declaration
 static void dmamux_stm32_irq_handler(arg);

erwango reviewed (changes requested):
First batch of comments

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -380,25 +545,45 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
-	for (;;) {
-		if (!stm32_dma_disable_stream(dma, id)) {
+#endif /* CONFIG_DMA_STM32_V1 */
+#if defined(CONFIG_DMA_STM32_V2) && defined(CONFIG_DMAMUX_STM32)
```
#ifdef CONFIG_DMAMUX_STM32

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -518,6 +715,7 @@ static const struct dma_driver_api dma_funcs = {
 	.config		 = dma_stm32_configure,
 	.start		 = dma_stm32_start,
 	.stop		 = dma_stm32_stop,
+	.get_status  = NULL,
```
Add a valid function that returns -ENOSYS.
To be done in another commit

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -589,20 +790,109 @@ static void dma_stm32_config_irq_1(struct device *dev)
+	return 0;
+}
+
+static const struct dma_driver_api dmamux_funcs = {
```
If API is not supported, device should be instantiated with DEVICE_INIT.

erwango replied:
Can you laborate on the reason you have to do this ?

FRASTM replied:
The `static const struct dma_driver_api dmamux_funcs = dmamux `
has no api functions by itself, 
but requires a declaration of NULL functions for the DEVICE_API_AND_INIT. 
If using DEVICE_INIT, will make the return a NULL pointer of the `device_get_binding((const char *)"DMAMUX_1");`

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -589,20 +790,109 @@ static void dma_stm32_config_irq_1(struct device *dev)
+	}								\
+} while (0)
+
+DMAMUX_INIT(0);
```
If only one instance, please avoid preprocessor macros for declaration.

erwango commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -316,7 +316,6 @@
 			reg = <0x40020000 0x400>;
 			interrupts = <11 0 12 0 13 0 14 0 15 0 16 0 17 0>;
 			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x1>;
-			st,mem2mem;
```
Is that change valid for whole series?

erwango commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -327,7 +326,6 @@
 			reg = <0x40020400 0x400>;
 			interrupts = <56 0 57 0 58 0 59 0 60 0 68 0 69 0>;
 			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x2>;
-			st,mem2mem;
```
ditto

FRASTM replied:
removed

erwango commented on tests/drivers/dma/mux_req_gen/prj.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_TEST=y
+CONFIG_DMA=y
+CONFIG_DMAMUX_STM32=y
```
This should not be needed.
Symbol DMAMUX_STM32 should be enabled if DMA is true on SoCs supporting it

FRASTM replied:
removed

erwango commented on tests/drivers/dma/mux_req_gen/prj.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_DMAMUX_STM32=y
+CONFIG_LOG=y
+CONFIG_DMA_LOG_LEVEL_INF=y
+CONFIG_EXTI_STM32=y
```
This should not be required (always 'y')

FRASTM replied:
removed

erwango commented on tests/drivers/dma/mux_req_gen/src/dma.c:
```diff
@@ -0,0 +1,159 @@
+/* dma.c - DMA test source file */
+
+/*
+ * Copyright (c) 2016 Intel Corporation.
```
Use your own copyright

FRASTM replied:
ok

erwango commented on tests/drivers/dma/mux_req_gen/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
I think this would be better as a sample

FRASTM replied:
changed but this application is too specific to commit

FRASTM commented:
the definition of the st,mem2mem is for DMA V1
to set if the DMA controller V1 supports memory to memory transfer
this is the case for DMA 2 of stm32f2, f4, f7

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -359,7 +494,36 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 	}
-#endif
+#endif /* CONFIG_DMA_STM32_V1 */
+#if defined(CONFIG_DMA_STM32_V2) && defined(CONFIG_DMAMUX_STM32)
```
fifo control exists for DMA V1

FRASTM commented:
splitting the PR in 2 parts  #22100 plus this one   #22021

FRASTM commented:
fix Nits issues

FRASTM commented:
for devices of the stm32l4 serie, some soc have dmamux.
(L45x2 has no dmamux, L4r5 has dmamux)
Nb of dma channels may also changed (ex. 8 for L5)

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -220,18 +358,15 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+/* In case support_m2m is true, only mem2mem transfer is allowed */
+#if !defined(CONFIG_DMAMUX_STM32)
+	if ((stream->direction != MEMORY_TO_MEMORY) &&
+		(dev_config->support_m2m)) {
```
moved

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -36,6 +99,24 @@ static void dma_stm32_clear_stream_irq(struct device *dev, u32_t id)
 	func_ll_clear_ht[id](dma);
 	stm32_dma_clear_stream_irq(dma, id);
+
+#if defined(CONFIG_DMA_STM32_V2) && defined(CONFIG_DMAMUX_STM32)
```
#if defined(CONFIG_DMAMUX_STM32)

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -199,6 +307,36 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+					dev_mux->config->config_info;
+	/* retrieve the dma channel from the dmamux channel */
+	u32_t mux_id = ((strcmp(dev->config->name, "DMA_1") == 0) ? id
+	 : ((devmux_config->channel_nb / 2) + id));
```
see dma_stm32.h
	/*
	 * macro to Retrieve the dmamux channel from the dma instance and channel (stream)
	 * DMAMUX channel 0 to channel_nb/2 -1 for DMA 1 channel (stream) id
	 * DMAMUX channel channel_nb/2 to channel_nb for DMA 2 channel (stream) id
	 */

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -23,10 +23,17 @@ config DMA_STM32_V1
 config DMA_STM32_V2
 	bool
-	depends on SOC_SERIES_STM32F0X || SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32L0X || SOC_SERIES_STM32L4X ||SOC_SERIES_STM32WBX
+	depends on SOC_SERIES_STM32F0X || SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32L0X || SOC_SERIES_STM32L4X ||SOC_SERIES_STM32WBX || SOC_SERIES_STM32G4X
```
removed

FRASTM commented:
Fixing the error when using another board like  96b_argonkey.
- Remove the `dma_stm32_config` parameter `dma_channels` as the nb of channels is given by the nb of irq through the `max_streams` of the `dma_stm32_data`
- for all the series stm32xx with dma capability, set the `HEAP_MEM_POOL_SIZE` to 1024 when the DMA is enabled

FRASTM commented:
rebase

FRASTM commented:
This PR enables the dmamux feature on the stm32 soc family
It configures the dma and mux feature on the stm32l4R5 soc series 
and use the board nucleo_l4r5zi to test dma mem-to-mem transfer with the test applications
It requires the https://github.com/zephyrproject-rtos/zephyr/pull/22100

FRASTM commented:
do not consider stm32wb series yet

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -447,6 +638,18 @@ static int dma_stm32_start(struct device *dev, u32_t id)
+#ifdef CONFIG_DMAMUX_STM32
+	/*
+	 * when using the dmamux, in periph2memm or mem2periph
+	 * counter and destination address counter.
```
Comment is not crystal clear

FRASTM replied:
updated

erwango commented on modules/Kconfig.stm32:
```diff
@@ -558,6 +558,12 @@ config USE_STM32_LL_DMA2D
 	  Enable STM32Cube Chrom-Art Accelerator™ controller (DMA2D) LL module
 	  driver
 
+config USE_STM32_LL_DMAMUX
```
`USE_STM32_LL_FOO` are only defined for ll*.c files.
This does not exist in case of dmamux.

FRASTM replied:
static const struct dma_driver_api dmamux_funcs = 
dmamux has no api functions by itself, but requires a declaration of NULL functions for the DEVICE_API_AND_INIT. If using DEVICE_INIT, will make the return a NULL pointer of the device_get_binding((const char *)"DMAMUX_1");

FRASTM replied:
ok, removed from modules/Kconfig.stm32

erwango commented on soc/arm/st_stm32/stm32l4/dts_fixup.h:
```diff
@@ -335,4 +335,24 @@
+#define DT_DMAMUX_1_IRQ_PRI	DT_ST_STM32_DMAMUX_40020800_IRQ_0_PRIORITY
+#define DT_DMAMUX_1_NAME		DT_ST_STM32_DMAMUX_40020800_LABEL
+#define DT_DMAMUX_1_CLOCK_BITS	DT_ST_STM32_DMAMUX_40020800_CLOCK_BITS_0
+#define DT_DMAMUX_1_CLOCK_BUS	DT_ST_STM32_DMAMUX_40020800_CLOCK_BUS_0
```
These are not needed since DMA instanciation code uses DT_INSTX_FOO defines

FRASTM replied:
removed

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -27,4 +27,12 @@ config DMA_STM32_V2
+	bool
+	depends on DMA_STM32_V2
+	depends on SOC_STM32L4R5XX
+	select USE_STM32_LL_DMAMUX
```
This is not needed.

FRASTM commented:
do not consider stm32wb series yet

commit b167981484f22039a0599a1a81a534230589a1c7 comes from boards: arm: add dma test support for some nucleo boards #20543

FRASTM commented:
rebase

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -605,14 +605,17 @@ static void dma_stm32_config_irq_0(struct device *dev)
+#endif /* DT_INST_0_ST_STM32_DMA_IRQ_5 */
+#endif /* DT_INST_0_ST_STM32_DMA_IRQ_6 */
+#endif /* DT_INST_0_ST_STM32_DMA_IRQ_7 */
+/* Either 5 or 6 or 7 or 8 channels for DMA1 across all stm32 series. */
```
removed

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -621,20 +624,20 @@ static void dma_stm32_config_irq_1(struct device *dev)
+#endif /* DT_INST_1_ST_STM32_DMA_IRQ_5 */
+#endif /* DT_INST_1_ST_STM32_DMA_IRQ_6 */
+#endif /* DT_INST_1_ST_STM32_DMA_IRQ_7 */
+/* Either 0 or 5 or 6 or 7 or 8 channels for DMA2 across all stm32 series. */
```
removed

FRASTM commented on dts/bindings/dma/dmamux-controller.yaml:
```diff
@@ -0,0 +1,37 @@
+    req-nb:
+      type: int
+      required: true
+      description: Number of DMAMUX Peripheral Request inputs supported by the controller
```
these parameters of the DMA are used to control the configuration of the client

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -27,4 +27,12 @@ config DMA_STM32_V2
 	help
```
fixed

FRASTM commented:
Checkpatch : fix warning with commit d34df336b17414e091a01e1bb827de4c05f1b9fd

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -28,15 +28,16 @@ static void dmamux_stm32_irq_handler(void *arg)
 	struct device *dev = arg;
 	struct dmamux_stm32_data *data = dev->driver_data;
 	const struct dmamux_stm32_config *config = dev->config->config_info;
-	DMAMUX_Channel_TypeDef *dmamux = (DMAMUX_Channel_TypeDef *)(config->base);
```
Move these fixes directly in faulty commit.

FRASTM replied:
done

FRASTM commented:
Move these fixes (line length warning) directly in faulty commit.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -379,25 +565,53 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+#ifdef CONFIG_DMAMUX_STM32
+	if (config->channel_direction != MEMORY_TO_MEMORY) {
+		/* there is also a DMAMUX channel synchro to set */
+		if (mux_channel->sync_en) {
```
I don't understand who is setting this.

FRASTM replied:
this is the setting of the request line mux configuration register

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -358,7 +504,46 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+			/* force LL_DMAMUX_REQ_GEN_POL_RISING */
+			mux_channel->gen_pol = 0x01;
+			mux_channel->gnb_req = 1;
+			LOG_ERR("       with fixed configuration.");
```
This is the first step before configuring from the client dts

FRASTM commented on drivers/dma/dma_stm32.c:
```diff
@@ -358,7 +504,46 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+		/* dma_slot corresponds to the periph request to set */
+		if (config->dma_slot == 0) {
+		/* a value of 0 means MEMORY_TO_MEMORY */
+			LOG_ERR("dmamux periph request cannot be NULL.");
```
ok

FRASTM commented:
more simple version which enables the dmamux with basic transfer

FRASTM commented on drivers/dma/dma_stm32.h:
```diff
@@ -22,6 +22,76 @@ static u32_t table_p_size[] = {
+	 * DMAMUX channel 0 to channel_nb/2 -1 for DMA 1 channel (stream) id
+	 * DMAMUX channel channel_nb/2 to channel_nb for DMA 2 channel id
+	 */
+#define get_dmamux_channel_from_dma(__chan)	\
```
changed with `__stream`

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -203,6 +249,49 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
+	 * Retrieve the dmamux channel from the dma instance
+	 * and channel (stream)
+	 */
+	u32_t mux_id = get_dmamux_channel_from_dma(id);
```
Now it is known, can we store `mux_id` in a dedicated `dma_stm32_data` field ?

FRASTM replied:
done,
adding a field `mux_id` to the `dma_stm32_stream` structure

FRASTM commented:
add one commit c8590afb83ef62a7a1f99ef5bc3c8ebe852c23cf to fix new error from shippable

erwango commented on drivers/dma/dma_stm32.h:
```diff
@@ -82,7 +85,6 @@ struct dma_stm32_config {
 	void (*config_irq)(struct device *dev);
 	bool support_m2m;
 	u32_t base;
-	u8_t dma_channels; /* the total nb of channels for this DMA */
```
To be done in another commit

FRASTM replied:
not used anymore

erwango commented on drivers/dma/dma_stm32.h:
```diff
@@ -63,6 +63,9 @@ struct dmamux_stm32_config {
 	u32_t direction;
+#ifdef CONFIG_DMAMUX_STM32
+	int mux_id;
+#endif /* CONFIG_DMAMUX_STM32 */
```
Can we try add dev_dmamux ?

erwango commented on drivers/i2s/Kconfig.stm32:
```diff
@@ -5,7 +5,7 @@
 menuconfig I2S_STM32
 	bool "STM32 MCU I2S controller driver"
-	depends on SOC_SERIES_STM32F4X
+	depends on SOC_SERIES_STM32F4X || SOC_SERIES_STM32G0X || SOC_SERIES_STM32G4X
```
To be cleaned up

FRASTM commented:
rebase and fix warning

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -38,6 +38,29 @@ static void dma_stm32_clear_stream_irq(struct device *dev, u32_t id)
+	if (dev_mux == NULL) {
+		LOG_ERR("cannot get dmamux device.");
+		return;
+	}
```
I think this could be removed now.

FRASTM replied:
removed

erwango reviewed (approved):
checkpatch warning seems wrong

galak reviewed (changes requested):
Seems like the DT bits need a bit of work:

1. Align w/linux binding:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/dma/stm32-dmamux.txt

2. Seems like either dmamux-controller.yaml or st,stm32-dmamux.yaml should include dma-controller.yaml

3. Bit confused on which side is the DMA controller and which to the device/master using the DMA MUX?

galak commented on dts/arm/st/l4/stm32l4r5.dtsi:
```diff
@@ -212,6 +212,19 @@
 
+		dmamux1: dmamux@40020800 {
+			compatible = "st,stm32-dmamux";
+			#dmamux-cells = <4>;
```
where is #dmamux-cells used?  Don't see anything in the bindings.

FRASTM replied:
now changed based on the dmamux bindings from linux

galak reviewed (commented):
Would also be good to add a sample or test showing how dmamux works.

FRASTM commented:
rebase

FRASTM commented:
Updates with redisgned based on the linux bindings
[https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/dma/stm32-dmamux.txt](url)

All dma controls from application are routed through the dmamux to the corresponding dma stream.
The dmamux api are in  a dedicated dmamux_stm32 area

FRASTM commented:
use dmamux function  LL_DMAMUX_SetRequestID in the dmamux_stm32

erwango commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -316,6 +316,7 @@
 			reg = <0x40020000 0x400>;
 			interrupts = <11 0 12 0 13 0 14 0 15 0 16 0 17 0>;
 			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x1>;
+			st,mem2mem;
```
Does it apply to v2?

FRASTM replied:
yes, meaning is defined by linux: support of mem2mem

erwango commented on dts/bindings/dma/st,stm32-dmamux.yaml:
```diff
@@ -0,0 +1,31 @@
+
+# Parameter syntax of stm32 follows the dma client dts syntax
+# in the Linux kernel declared in
+# https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/stm32-dmamux.txt
```
Link fails

FRASTM replied:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/st,stm32-dmamux.yaml
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/st,stm32-dma.yaml

FRASTM commented:
function to clear the global interrupt is for DMA_V2 only (func_ll_clear_gi)

FRASTM commented:
rebase

FRASTM commented:
with DMA_STM32 V2, handling the irq based on the global interrupt (gi)
Only transfer complete (tc) or error (te) irq have  callback function call

FRASTM commented:
DTS : bindings: put the link to linux kernel  st,stm32-dma.yaml and  st,stm32-dmamux.yaml

FRASTM commented:
Change the dma and dma mux configuration in the dts : 
- the channel is set in the dma-cell for dmamux or dma
- the slot is for dmamux peripheral request or dma peripheral request in dma V2 or dma slot in dma V1
When the dma mux exists, the dmamux API calls the dma API on the corresponding dma channel
The dma interrupt on a stream is directly routing the corrsponding dmamux channel to the dma_callback.
There is no dmamux specific irq handler.

FRASTM commented:
remove the channel nb parameter in the dma_stm32_config and use the data->max_streams instead,  in the dma_stm32_init to assign the corresponding dmamux channel to dma channel

FRASTM commented:
this PR adds the dma feature on top of the spi_dma PR https://github.com/zephyrproject-rtos/zephyr/pull/23157

FRASTM committed:
dts: bindings: dma add support for dmamux controller

This patch introduces the periph to/from memory dma transfer
using the dmamux feature and define new values
for dma cells on client side

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase

tbursztyka commented on drivers/dma/dma_stm32.c:
```diff
@@ -93,24 +93,45 @@ static void dma_stm32_irq_handler(void *arg)
 	}
 
 	stream = &data->streams[id];
+#ifndef CONFIG_DMAMUX_STM32
```
would be nicer if you used if (IS_ENABLED()) in relevant places  (not possible for mux_channel attribute though)

FRASTM replied:
changed in dma_stm32_reload and in dma_stm32_reload functions

FRASTM commented:
rebase

FRASTM commented:
using `if (IS_ENABLED(CONFIG_DMAMUX_STM32))` in dma_stm32_reload and in dma_stm32_reload functions

FRASTM commented:
rebase after the merge operation of the https://github.com/zephyrproject-rtos/zephyr/pull/24070
This PR adds the dmamux drivers on stm32.

FRASTM commented:
Always disable/enable the dma channel during the dma reload operation. This makes the transfer restart with new source/destination addresses.

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -516,7 +570,12 @@ static int dma_stm32_stop(struct device *dev, u32_t id)
 
+#ifdef CONFIG_DMAMUX_STM32
+	/* do not disable TC irq */
+#else
```
Why not doing ifndef ?

FRASTM replied:
This comment I want to add in case of dmamux

avisconti replied:
Maybe you can do:

--
/* do not disable TC irq in case of DMAMUX */
#ifndef CONFIG_DMAMUX_STM32
...
#endif

FRASTM replied:
No, this comment is meaningless without DMAMUx

erwango replied:
Once again, I prefer avoiding empty sections.

avisconti replied:
> Once again, I prefer avoiding empty sections.

@erwango 
I agree

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -519,9 +516,7 @@ static int dma_stm32_reload(struct device *dev, u32_t id,
+	stm32_dma_enable_stream(dma, id);
 
 	return 0;
 }
```
Shouldn't this commit be squashed with the previous? 
Isn't the file the same?

FRASTM replied:
Yes, but I did in 2 commits, but yes one change is useless

avisconti replied:
So, did you squash it at the end?

avisconti commented on drivers/dma/dmamux_stm32.c:
```diff
@@ -0,0 +1,291 @@
+		if (func_ll_is_active_so[i](dmamux)) {
+			break;
+		}
+		if (stm32_dmamux_is_irq_happened(dmamux, i)) {
```
Maybe you can or this with the precedent

FRASTM replied:
yes
changed

avisconti replied:
Apparently this code has been replaced by something else.
Can you possibly explain to me what happened?

avisconti replied:
@erwango 
Can you confirm the code has changed here?

erwango replied:
@avisconti Yes, this was actually dead code, since dmamux irq is not used (it's only purpose for now is to be used as proxy), we don't need irq handling for now.

avisconti commented:
@FRASTM 
Just for a better understanding. DMAMUX is a peripheral that is embedded in some stm32 SoC, correct? If so, can you mention it in the PR header itself? Moreover it seems to me that this peripheral is used to multiplex requests coming from other peripherals to more different dma controllers in the SoC itself. Is it the case?

FRASTM committed:
drivers: dma: Enable dma and dmamux on stm32

This patch prepares the dma and introduces the dmamux on soc series
which supports this feature for memory/periph transfers.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
change the dmamux driver to use the DT macro

erwango reviewed (changes requested):
Please squash:
drivers: dma: Enable dmamux on stm32 for memory/periph modes
with 
drivers: dma: Enable dma to support dmamux on stm32

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -286,6 +313,9 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 	stream->callback_arg    = config->callback_arg;
 	stream->src_size	= config->source_data_size;
 	stream->dst_size	= config->dest_data_size;
+#ifdef CONFIG_DMAMUX_STM32
+	/* the stream->mux_channel is fixed by the dma init */
+#endif /* CONFIG_DMAMUX_STM32 */
```
I'm doubtful about comments under compile flags. And potentially some static code analyzer may complain.
Can you just reword this part to make clear that incase of DMAMUX stream->mux_channel is fixed by the dma init ?

FRASTM replied:
moving this comment directly in the dma_init function

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -22,9 +22,23 @@ config DMA_STM32_V1
+	help
+	  Enable DMAMUX support on L4R/WB series SoCs.
+
+config DMAMUX_1_NAME
+	string "Device name for DMAMUX Controller 1"
+	depends on DMAMUX_STM32
+	default "DMAMUX_1"
+	help
+	  Device name for DMAMUX Controller 1.
```
Is that needed ?
I guess this should not be if using DT macros

FRASTM replied:
yes, removed provided through the DT

erwango reviewed (changes requested):
Minor comments otherwise LGTM

FRASTM commented:
reduce the Kconfig for dmamux name

erwango commented on drivers/dma/dmamux_stm32.c:
```diff
@@ -0,0 +1,219 @@
+	 */
+
+	/* give arg for the dmamux callback function */
+	data->callback_arg = (void *)dev;
```
Is that used ?

FRASTM replied:
no, removed

erwango commented on dts/bindings/dma/dmamux-controller.yaml:
```diff
@@ -0,0 +1,36 @@
+      required: false
+      description: Number of items to expect in a DMA specifier
+
+    dma-masters:
```
Remove if not used

FRASTM replied:
removed

erwango commented on dts/bindings/dma/dmamux-controller.yaml:
```diff
@@ -0,0 +1,36 @@
+      required: true
+      description: Number of DMAMUX output request channels supported by the controller
+
+    dma-generators:
```
removed if not used

FRASTM replied:
yes, for stm32wb55 the request ID cannot exceed 40 = Number of DMAMUX peripheral request inputs + Number of DMAMUX request generator channels
no, for stm32g4 the request ID cannot exceed 115 =  Number of DMAMUX peripheral request inputs

erwango replied:
problem is that I can't see where it is used..

FRASTM replied:
dmamux gen_nb  is compared to the given request ID : request ID must not exceed Number of DMAMUX request generator channels + Number of DMAMUX peripheral request inputs (for stm32wb

erwango commented on dts/bindings/dma/st,stm32-dmamux.yaml:
```diff
@@ -0,0 +1,86 @@
+         compatible = "st,stm32-dmamux";
+         ...
+         dma-channels = <14>;
+         dma-generators = <4>;
```
To be cleaned if removed

FRASTM replied:
removed

FRASTM commented:
- squashing
`drivers: dma: Enable dmamux on stm32 for memory/periph modes`
with
`drivers: dma: Enable dma to support dmamux on stm32`

- clean dts/bindings

galak commented on dts/bindings/dma/dmamux-controller.yaml:
```diff
@@ -0,0 +1,32 @@
+
+    "#dma-cells":
+      type: int
+      required: false
```
why is this false?

FRASTM replied:
none, 
setting as true like in dma-controller.yaml

avisconti reviewed (approved):
All my 3 points raised have been addressed.

FRASTM commented:
rebase
